File: question/neetcode150/stack/largest-rectangle-in-histogram.py
from typing import List

class Solution:
    def largestRectangleArea(self, a: List[int]) -> int:
        stack = []
        ans = 0
        for i,n in enumerate(a):
            start = i
            while stack and stack[-1][1] > n:
                start, height = stack.pop()
                ans = max(ans, height * (i-start))

            stack.append((start, n))

        for start, height in stack:
            ans = max(ans, height * (len(a)-start))
        
        return ans

File: question/neetcode150/stack/largest-rectangle-in-histogram.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int largestRectangleArea(vector<int>& a) {
        int ans = 0;
        int n = a.size();
        stack<pair<int,int>> s;

        for(int i = 0;i < n;i++){
            int h = a[i];
            int k = i;

            while(not s.empty() and s.top().second > h){
                auto [j,v] = s.top();
                s.pop();
                ans = max(ans, (i-j)*v);
                k = j;
            }

            s.push({k,h});
        }

        while(not s.empty()){
            auto [i,h] = s.top();
            s.pop();
            ans = max(ans, (n-i)*h);
        }

        return ans;
    }
};